# try-finally 보다는 try-with-resources 를 사용하라

## try-finally 사용의 문제점

* 자바 라이브러리에는 ```close```메서드를 호출해 직접 닫아줘야 하는 자원이 많다.
* ```InputStream, OutputStream, java.sql.Connection``` 등이 좋은 예다.
* 이러한 자원 닫기는 클라이언트에서 놓치기 쉬워 예측할 수 없는 성능 문제로 이어지기도 한다.
  이러한 자원 중 상당수에서 안전망으로 ```finalizer```를 활용하고 있지만 [아이템 8](https://github.com/parkhanbeen/study/blob/master/effective-java/2%EC%9E%A5/8.finalizer%EC%99%80%20cleaner%20%EC%82%AC%EC%9A%A9%EC%9D%84%20%ED%94%BC%ED%95%98%EB%9D%BC.md)
  에서 나온것 처럼 ```finalizer```는 믿을만하지 못하다.
* 보통 자원을 제대로 닫기 위해서 ```try-finally```가 쓰인다.
* 하나의 자원만 닫을 경우
```java
static String firstLineOfFile(String path) throws IOException {
  BufferedReader br = new BufferedReader(new FIleReader(path));
  try {
    return br.readLine();
  } finally {
    br.close();
  }
        
}
```
* 둘 이상의 자원을 닫을 경우
```java
static void copy(String src, String dst) throws IOException {
  InputStream is = new FileInputStream(src);
  try {
    OutputStream out = new FileInputStream(src);
    try {
      byte[] buf = new byte[BUFFER_SIZE];
      int n;
      while ((n = in.read(buf))) >= 0)
        out.write(buf, 0, n);
    } finally {
      out.close();
    }
  } finally {
    in.close();
  }
}
```
* 둘 이상의 자원을 닫을 경우 소스가 지저분해진다.
* 예외는 ```try``` 불록과 ```finally``` 블록 모두에서 발행할 수 있는데, 예컨대 기기에 물리적인 문제가
  생긴다면 ```firstLineOfFile``` 메서드 안의 ```readLine``` 메서드가 예외를 던지고 같은 이유로 ```close```
  메서드도 실패할 것이다.
* 이러한 상황은 스택 추적 내역에 첫 번째 예외에 관한 정보는 남지 않게 되어, 실제 시스템에서의 디버깅을 몹시 어렵게 한다.

## try-with-resources 를 이용하여 해결

* 위의 ```try-finally```의 문제점을 자바7에서 나온 ```try-with-resources```로 해결할 수 있게 되었다.
* ```try-with-resources``` 구조를 사용하려면 ```AutoCloseable```인터페이스를 구현해야 한다.
  * [AutoCloseable 공식 문서](https://docs.oracle.com/javase/7/docs/api/java/lang/AutoCloseable.html) 
  * ```AutoCloseable```인터페이스는 단순히 ```void``` 리턴 타입을 반환하는 ```close```메서드 하나만 명시된 인터페이스이다.
  * 여러 자바 라이브러리와 서드파티 라이브러리들의 수많은 클래스와 인터페이스가 이미 ```AutoCloseable```인터페이스를
    구현하거나 확장하였다.
```java
public interface AutoCloseable {
  void close() throws Exception;
}
```
* ```try-with-resources```를 하나의 자원만 닫을 경우
```java
static String firstLineOfFile(String path) throws IOException {
  try (BufferedReader br = new BufferedReader(new FileReader(path))) {
    return br.reaLine();
  }
}
```

* ```try-with-resources```를 둘 이상의 자원을 닫을 경우
```java
static void copy(String src, String dst) throws IOException {
  try (InputStream in = new FileInputStream(src);
        OutputStream out = new FileOutputStream(dst)) {
    byte[] buf = new byte[BUFFER_SIZE];
    int n;
    while ((n = in.read(buf)) >= 0) {
      out.write(buf, 0, n);
     }
   }
}
```
* ```try-with-resources```를 사용했을 경우가 코드도 더 짧아지고 읽기 수월할 뿐만 아니라 문제를 
  진단하기도 휠씬 쉬워졌다.
* ```firstLineOfFile``` 메서드를 생각해보면 ```readLine```과 ```close``` 양쪽에서 예외가 밣생하면
  ```close```에서 발생하는 예외는 숨겨지고 ```readLine```에서 발생한 예외가 기록된다.
* 이처럼 숨겨진 예외는 버려지지 않고 **스택 추적 내역에 숨겨졌다는 꼬리표(suppressed)** 를 달고 출력된다.
* 또한, 자바7에서 ```Throwable```에 추가된 ```getSuppressed``` 메서드를 이용하면 프로그램 코드에서 가져올 수도 있다.
* 보통의 ```try-finally```처럼 ```catch``` 절을 사용할 수도 있다.
* ```catch``` 절 덕분에 ```try```문을 더 중첩하지 않고도 다수의 예외를 처리할 수 있다.
```java
static String firstLineOfFile(String path, String defaultVal) {
  try (BufferedReader br = new BufferedReader(new FileReaader(path))) {
    return br.readLine();
  } catch (IOException e) {
    return defaultVal;
  }
}
```

## 정리

* 꼭 회수해야 하는 자원을 다룰 때는 ```try-finally``` 말고, ```try-with-resources```를 사용하도록 하자